### DISCUSSION TITLE
Open form screen for specific row

### DISCUSSION THREAD

**Discussion by sstrasserJember** [Budibase/budibase#5501] [Help] [answered]
Hi all,

I'm pretty shure this is somewhere in the documentation or in this discussion board, but I can't find it and I'm starting to question my sanity.
I have a pretty similar request as https://github.com/Budibase/budibase/discussions/4156

I have a table with a lot of columns and I want to create several screens that only show parts of the table.
The reason behind is: We need an onboarding process, where HR only puts in stuff like the name, department and starting date of the employee. Then a mail is sent to the department lead so they can open the row with the already provided data shown as disabled and can then request hardware, software and so on.
What I just can't wrap my head around is:
How do I configure the Edit button on the card representing a row in the table to open my form screen for the update on that specific row?

My initial screen for the department leaders (overview) is set up like this:
![grafik](https://user-images.githubusercontent.com/104074591/164227985-61eda41e-8ce2-4371-8ca5-4a70f638fc66.png)
Where RequirementData is the data provider set to the table in question.
The provider already has a filter for binding _id to {{ URL.id }}, as specified in the link above.
The "update that row" screen is set up like this:
![grafik](https://user-images.githubusercontent.com/104074591/164228375-3ea3486b-4399-41c2-8622-2254b2c5989f.png)
Where User is a form and Info a field group.

For my form fields, I used text fields with a default value of {{ User.Fields.xxx }}, but that won't do the trick.
Also, even though I created the data provider as a parent, the table fields won't show up in the bindings helper:
![grafik](https://user-images.githubusercontent.com/104074591/164229293-7d6e60a7-9510-4e53-9c45-6d2de55f2790.png)

The Edit button in the overview screen is configured like this:
![grafik](https://user-images.githubusercontent.com/104074591/164229556-63791fd4-fb13-4f38-bbe8-b64cca47ebe2.png)

I'm at my wits end here. What am I missing? Where in the documentation is a case like this, as I'm pretty shure that it's a very common use of BudiBase.
To be honest, this is my very first time working with any kind of database, so consider me a noob and explain like I'm five.

**Comment by melohagan** [comment]
Hi @sstrasserJember 

I think the last screenshot shows the issue. Rather than having **:id** you want to have the binding in its place which would be **{{ URL.id}}**

So *Navigate To* `/info/update/row/{{ URL.id }}`

If you click the lightning bolt icon you should be able to see that option listed on the left hand panel.

Hope this helps!

**Reply by sstrasserJember** [reply]
Hi @melohagan !
Thanks for the info, but sadly, it's still not working.
![grafik](https://user-images.githubusercontent.com/104074591/164241462-3e24427b-df74-4a45-b68a-5b58e2c54168.png)
As you can see, the URL parameter is not shown in the helper box.
Also, I changed the Navigate To setting as you said and now the edit button does nothing anymore 😄 
![grafik](https://user-images.githubusercontent.com/104074591/164241732-0deb4675-4aa6-4fe3-b268-d5a2cb77b201.png)

**Reply by sstrasserJember** [reply]
I've even tried adding the row ID to the /info/update/row/ URL, like this:
/info/update/row/ro_ta_a070bdf6bb594b67bc93d783575c4fff_171a1eeebe1c441b829f30f6ce60830f

That only sets me back to /info

**Reply by melohagan** [reply]
Ah okay yes the overview screen doesn't have an id in the URL so that does make sense. 

You do still need to pass an id in the Navigate To URL. 

So for example if you have a table of department leaders in the overview screen, then you can pass in the row id. 
E.g. https://github.com/Budibase/budibase/discussions/1723#discussioncomment-881738

Basically if you are using an ID then your action needs to be on a per row basis. Then you'll have access to the _id of the row.

**Reply by melohagan** [reply]
This discussion may help as well: https://github.com/Budibase/budibase/discussions/5231

**Reply by sstrasserJember** [reply]
Uhm... my brain is fried 😄 I don't get it.
What do I need to change to make it work?
I have a Card Block in the overview screen that displays the rows from the table Userinfo. Each row represents a user.
The URL needs to be added but I still don't know how to add it in a format that makes sense for BudiBase x_x
As I said, consider me a 5yo 😆

**Reply by sstrasserJember** [reply]
Hm, when I changed the linked title setting to /info/update/row/:id, the update screen opens with a correct ID. But I still can't access the corresponding fields:
![grafik](https://user-images.githubusercontent.com/104074591/164248171-422d68be-02a1-4465-ba11-abc8a073fdfa.png)
Produces
![grafik](https://user-images.githubusercontent.com/104074591/164248212-3720818e-262e-45c5-a900-774134b598fd.png)

Config of the update screen is
![grafik](https://user-images.githubusercontent.com/104074591/164248329-8d59a2c6-7382-44b6-8d4d-427b2632004f.png)

If I can fill the fields with the correct data, I'm good 😆 

To clarify: The behaviour is only correct, when I click the linked title. The Edit button is still kinda useless, but I can use it differently, so that's not an issue.

**Reply by melohagan** [reply]
Huh that is strange.

Would you be able to export your app (if there's no personal info in it) ? I do need to update the docs to go more into URL params, so apologies for the lack of clarity!

**Reply by sstrasserJember** [reply]
[UserOnboarding-export-1650463982877.txt](https://github.com/Budibase/budibase/files/8522216/UserOnboarding-export-1650463982877.txt)

Sure, here you go ;)
Nothing that's really important in there.

**Comment by melohagan** [comment, marked as answer]
Okay @sstrasserJember, I believe I've got it working, and I'll try to explain as best I can how it works; and use this as feedback for the docs :)

### The Difference between Links and Navigate To Action

There are two approaches to linking screens with an ID. 

The first using the 'Navigate To' action:
![Screenshot 2022-04-20 at 15 37 21](https://user-images.githubusercontent.com/101575380/164255674-594bef31-e36f-49eb-ba07-8c64073260cb.png)
*In this case each card block has a useable binding to the id of that row*

The second using the Link:
![Screenshot 2022-04-20 at 15 36 07](https://user-images.githubusercontent.com/101575380/164255362-b8d1b9d2-90d6-483a-84d9-74bb428ef50b.png)
*In this case, because it's a link, there is a magic binding that automatically knows to grab the id used above*

### Default form values for a given URL ID

1. Add a data provider that uses the table of interest as it's source:
   ![Screenshot 2022-04-20 at 15 46 12](https://user-images.githubusercontent.com/101575380/164257692-f165eeab-ca29-48f6-babb-cfea7a10edb6.png)

2. Click 'Define Filters'. We only want the data for the user as specified by the id passed in the URL:
    ![Screenshot 2022-04-20 at 15 47 21](https://user-images.githubusercontent.com/101575380/164257924-84e1a9f6-9a43-412f-8abc-566ccc4f702c.png)

3. Add a repeater to your data provider, and place your form within the repeater:
    | Settings | Tree |
    | -------- | ---- |
   | ![Screenshot 2022-04-20 at 15 48 22](https://user-images.githubusercontent.com/101575380/164258136-a528d3e7-8cb2-4626-a67c-c7c22b628a07.png) | ![Screenshot 2022-04-20 at 15 49 57](https://user-images.githubusercontent.com/101575380/164258454-551dc886-a839-458c-804a-1083b42ec64f.png) |

4. For the default values of your form fields, use the bindings from the repeater:
    ![Screenshot 2022-04-20 at 15 52 25](https://user-images.githubusercontent.com/101575380/164259051-b590dfbc-4264-4ab4-b7cf-edc913abbc8a.png)

5. That should be it working! You will see in the designer a list of repeated forms for each user. No need to worry as the filter we set up will ensure only one user is shown in the app.

----
Here is the app export so you can have a look: [Passing ID-export-1650466503383.txt](https://github.com/Budibase/budibase/files/8522615/Passing.ID-export-1650466503383.txt)

As a side note, if you are collecting form data in steps, then this may be handy: https://docs.budibase.com/docs/form-steps

Also, I saw you were using bindings from the User table. This table is the built in Budibase table that keeps track of users that log into the Budibase app. You can add custom columns to this table, and create relationships between it and other tables, but that's entirely optional.

**Reply by sstrasserJember** [reply]
Hi @melohagan !
That is way more clear than what's in the documentation, thank you so much! I'll try implementing it later, so I can give you more feedback.
It would be great if the documentation contained more detailed info on the variables and how to access data. For example, I still don't know why the Card Block is used as a data provider, or why I need a data provider in the first place, instead of just accessing the table directly.
Also, why exactly do I need the repeater inside the data provider? Is it to iterate through the table?

I already looked at the form steps, but the thing is that the workflow I'm building is time-asynchronous. Between step 1 and step 2, several days can pass and an email needs to be sent to the user, because step 1 is done by HR and step 2 by the department head 🤷 

But still thanks for the awesome feedback! If I get my app to work properly and learn some more about the issues I had problems with, maybe I can help with the documentation.

**Reply by melohagan** [reply]
Your feedback is very welcome! :)
Off the back of this discussion, I did add this docs page which will hopefully help others in future: https://docs.budibase.com/docs/url-parameters

So the <a href="https://docs.budibase.com/docs/repeater">Repeater</a> loops over the rows of a data provider and repeats anything nested underneath it. So you could have cards, charts, form fields, etc, for every row.

The reason a repeater is used here is because we have a filter on the data provider. So essentially the data provider returns a list of rows, which because we filtered on the row-id will just return a list with a single row.
Then the repeater iterates over that list of one, thus showing the form data for just that user. 

I really like this recent idea to combine this into a single component, as it is used a lot: https://github.com/Budibase/budibase/discussions/1176#discussioncomment-2565544

Feel free to add ideas in there, or open up a discussion anytime!

**Reply by melohagan** [reply]
The Cards Block: https://docs.budibase.com/docs/blocks#cards-block is just a convenient combination of a data provider, repeater, and cards.

**Reply by sstrasserJember** [reply]
It works!
The designer is really weird now, as it always shows ALL rows of the table, but the actual app shows only the one in question.
What still doesn't work is the update of the row.
The Save Row command looks like this:
![grafik](https://user-images.githubusercontent.com/104074591/164425675-13df4ddd-f658-43e5-995a-0f56d077878a.png)

The Data Source is the UserInfo Repeater. Selecting the form Requirements itself produces an error and choosing "none" does the same: "Row saved" output, but nothing shows up in the row. x_x Do I need to provide which row I want to update?

**Reply by melohagan** [reply]
You'll want the Update button to be nested under the form rather than the repeater. The form contains the schema that is needed for the *Save Row* to work. 

![Screenshot 2022-04-21 at 10 56 35](https://user-images.githubusercontent.com/101575380/164430252-ca41e88d-af2e-4304-abe9-f660f51f469c.png)

That said, the behaviour with the repeater seems strange. Seems to be a bug that'll look into further. 

Also, as you have a few required fields in the UserInfo table, you'll need to provide those columns when you save the row:
![Screenshot 2022-04-21 at 10 53 42](https://user-images.githubusercontent.com/101575380/164429696-1674d023-3fc3-45f9-8967-89cd8facf853.png)

It would probably be nicer if required fields were only required for creation, and not updating an existing row.

**Reply by sstrasserJember** [reply]
This is the behaviour inside the designer, if it helps:
![grafik](https://user-images.githubusercontent.com/104074591/164432666-8ca3992e-637e-4db7-82d7-11511b8d02b8.png)
As you can see, the form just starts again below (First Name).

Concerning the button placement, the buttons are in fact nested in the form:
![grafik](https://user-images.githubusercontent.com/104074591/164433022-0f6400ce-23d2-4839-8fc7-e8ade25a095a.png)
I've added the required fields as you asked, but still the same result:
![grafik](https://user-images.githubusercontent.com/104074591/164433269-dd92721e-dadc-4148-ac62-25792d306309.png)

Also, is there a way to update / save only some columns of a row? I've seen it in the automation: if I want to update the row, I always have to add the value {{ trigger.row.Field }} for ALL columns for it to work.

**Reply by melohagan** [reply]
Change the data source to point to the User form, and that should do the trick. 
You're spot on that if you want to update only some rows, then you would need to add all of the rows unfortunately.

**Reply by sstrasserJember** [reply]
I did the following now:
I moved the field group Requirements so it's nested under the Form User. This makes sure that all forms access the same data from the Data Provider.
![grafik](https://user-images.githubusercontent.com/104074591/164440043-8d0c458e-358d-4b1f-a8b9-66e411321984.png)

When I set the data source for the row update to the User form, I get a bad request when saving. When I select the data provider or none, I can save, but the row isn't changed.

Could it be that the reason I can't update the row is the column config? i.e. the column "HardwareRequirements" is set as an options field, with several options. The form actually uses these options and I can select from them, but might it be that the return value doesn't support these options as input?

**Reply by melohagan** [reply]
You wouldn't mind sharing the Save Row action screenshot again?

**Reply by sstrasserJember** [reply]
Not at all!
This config produces a bad request:
![grafik](https://user-images.githubusercontent.com/104074591/164448644-1919e174-143d-4fb0-8e79-615bd703c7e6.png)
And this just straight up does nothing (even though is says "row saved"):
![grafik](https://user-images.githubusercontent.com/104074591/164449156-01b71f1f-094b-42f0-ad56-c532662ce50a.png)

If you want, I can send you the app again, if it helps. 
Please keep in mind that this is not yet in production, nor is it really time critical. So take your time. ;)

**Reply by melohagan** [reply]
Do you have company and starting date set in there as well? Those fields were required in your table schema. The top data source (User form) is correct.

**Reply by sstrasserJember** [reply]
Please excuse me for a moment.
*GOD DAMNIT!!!*
Okay, back with you... Sorry you had to deal with my stupidity. The Company field is required and I forgot to set it, which caused the bad request. It works now as intended!
Thank you soooo much! I fear I'd never have solved it without you!
I wish you'll never have to deal with a user like me again :D

**Reply by melohagan** [reply]
No worries, just glad that you're sorted :)

**Reply by melohagan** [reply]
One last thing, if you find it's doing a row creation instead of an update then add in the **_id** field as well:

![Screenshot 2022-04-21 at 14 40 16](https://user-images.githubusercontent.com/101575380/164470613-297fbdc9-c2e5-4ff2-b217-e152d7a26487.png)

**Reply by sstrasserJember** [reply]
Isn't that what the behaviour set in the Form properties is? There's a switch between "Create" and "Update".
But the additional info might come in handy for anyone else looking for answers.

**Reply by sstrasserJember** [reply]
Oh, I've just seen that you're right, it creates new rows :D Well, you answered the question before it was even asked ;)
